bench/ecdsa.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015-2023 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+#![allow(missing_docs)]
+
+use criterion::{black_box, criterion_group, criterion_main, Criterion};
+use ring::{
+    rand,
+    signature::{self, EcdsaKeyPair, EcdsaSigningAlgorithm, EcdsaVerificationAlgorithm, KeyPair},
+};
+
+static ALGORITHMS: &[(&str, &EcdsaSigningAlgorithm, &EcdsaVerificationAlgorithm)] = &[
+    (
+        "p256",
+        &signature::ECDSA_P256_SHA256_FIXED_SIGNING,
+        &signature::ECDSA_P256_SHA256_FIXED,
+    ),
+    (
+        "p384",
+        &signature::ECDSA_P384_SHA384_FIXED_SIGNING,
+        &signature::ECDSA_P384_SHA384_FIXED,
+    ),
+];
+
+fn sign(c: &mut Criterion) {
+    for (alg_name, alg, _) in ALGORITHMS {
+        c.bench_function(&format!("ecdsa_{alg_name}_sign"), |b| {
+            let rng = rand::SystemRandom::new();
+            let pkcs8_bytes = EcdsaKeyPair::generate_pkcs8(alg, &rng).unwrap();
+            let key_pair = EcdsaKeyPair::from_pkcs8(alg, pkcs8_bytes.as_ref(), &rng).unwrap();
+
+            b.iter(|| {
+                key_pair.sign(&rng, black_box(&[])).unwrap();
+            })
+        });
+    }
+}
+
+fn verify(c: &mut Criterion) {
+    for (alg_name, sign_alg, verify_alg) in ALGORITHMS {
+        c.bench_function(&format!("ecdsa_{alg_name}_verify"), |b| {
+            let rng = rand::SystemRandom::new();
+            let pkcs8_bytes = EcdsaKeyPair::generate_pkcs8(sign_alg, &rng).unwrap();
+            let key_pair = EcdsaKeyPair::from_pkcs8(sign_alg, pkcs8_bytes.as_ref(), &rng).unwrap();
+            let signature = key_pair.sign(&rng, &[]).unwrap();
```

Comment:
We're going to run into issue described in https://github.com/briansmith/ring/issues/1772 where the performance of each run is going to vary based on the randomized values of `r` and `s` in this signature.